freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RebootInstancesService.java

```diff
@@ -0,0 +1,116 @@
+package com.sequenceiq.freeipa.service.stack;
+
+import static com.sequenceiq.freeipa.flow.instance.reboot.RebootEvent.REBOOT_EVENT;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceStatus;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.health.NodeHealthDetails;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.reboot.RebootInstancesRequest;
+import com.sequenceiq.freeipa.controller.exception.BadRequestException;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.flow.instance.InstanceEvent;
+import com.sequenceiq.freeipa.service.freeipa.flow.FreeIpaFlowManager;
+import com.sequenceiq.freeipa.service.stack.instance.InstanceMetaDataService;
+
+@Service
+public class RebootInstancesService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RebootInstancesService.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaFlowManager flowManager;
+
+    @Inject
+    private FreeIpaHealthDetailsService healthDetailsService;
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    private Map<String, InstanceStatus> getInstanceHealthMap(String accountId, RebootInstancesRequest request) {
+        return healthDetailsService.getHealthDetails(request.getEnvironmentCrn(), accountId).getNodeHealthDetails().stream()
+                .collect(Collectors.toMap(NodeHealthDetails::getInstanceId, NodeHealthDetails::getStatus));
+    }
+
+    private Collection<String> getValidInstanceIds(Collection<String> allInstances, Collection<String> instanceIds) {
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            return allInstances;
+        }
+        boolean badIds = false;
+        Collection<String> validInstanceIds = new HashSet<>();
+        for (String instanceId: instanceIds) {
+            if (allInstances.contains(instanceId)) {
+                validInstanceIds.add(instanceId);
+            } else {
+                LOGGER.error("{} is an invalid instanceId.", instanceId);
+                badIds = true;
+            }
+        }
+        if (badIds) {
+            String msg = "Invalid instanceIds are present in request.";
+            LOGGER.error(msg);
+            throw new BadRequestException(msg);
+        }
```

Comment:
so we could say here `if(instanceIds.size() != validInstanceIds.size())` and just calculate the difference of the 2 collection and log at once and add it to the message.
this way the for could be a stream:
`Collection<String> validInstanceIds = instanceIds.stream().filter(allInstances.contains).collect()`
and you could drop the `badIds` flag

Reply:
changed.